### PR TITLE
Redis integration and code refactor

### PR DESCRIPTION
**Pull request by Timtech4u**
This PR contains the working code for Redis integration, allows saving valid responses to a Redis DB.

Work is in progress for adding validity check using Regex.

Closes issues #26  #27  #11

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -34,58 +17,101 @@ def ussd_handler():
...
         service_code = request_data['serviceCode']
         phone_number = request_data['phoneNumber']
-        text = request_data['text']
-        usrInput = text.split('*')
-        # Removes '' from initial usrInput List
-        while '' in usrInput:
-            usrInput.remove('')
-        # Gets the current USSD Stage
-        step = len(usrInput)
+        ussdTextList = request_data['text'].split('*')
```
Comment (tuxpiper) [inline app.py:24]: side note: did we ever test what would happen if the user entered text with asterisk symbols in it?

File: app.py [hunk 2]
```diff
@@ -34,58 +17,101 @@ def ussd_handler():
...
+        # Removes '' from list
+        while '' in ussdTextList:
```
Comment (tuxpiper) [inline app.py:27]: wouldn't filter() ( https://docs.python.org/3/library/functions.html#filter ) be a better approach?

Modifying a list while iterating on it is a source of concern for me (and others as well, i.e.: https://unspecified.wordpress.com/2009/02/12/thou-shalt-not-modify-a-list-during-iteration/ )

Comment (tuxpiper) [reply app.py:27]: Also, why would there be empty elements in the list and how removing them would affect step count?

File: app.py [hunk 3]
```diff
@@ -34,58 +17,101 @@ def ussd_handler():
...
+                    response += "\nEnter 0 to continue or cancel."
+                    # Initialize Redis with unique session_id
+                    redis_db.set(session_id, '')
+                    redis_db.append(session_id, form_id) # Save form ID
+                    redis_db.append(session_id, f"*{str(fields)}") # Save form Fields
+
+            return response
+
+        # Further USSD Interactions
+        if len(redis_db.get(session_id)) >= 1:
```
Comment (tuxpiper) [inline app.py]: is this equivalent to `validStep >=1` ?  If so, using that expression would make the flow a bit easier to understand while reading the code (imho)

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import os
+import ast
+
+from flask import Flask, request
+import redis
```
Comment (tuxpiper) [inline setup.py]: is there any corresponding entry for this module in requirements.txt ? I don't see one neither here on in master.

**Review by tuxpiper — COMMENTED** [review]
I'm finding it pretty hard to follow how the app is storing state in Redis. May I suggest creating a separate class/module that is solely responsible of formatting in and parsing out data from Redis?

It would really make the code cleaner, sp. if that class/module has methods/functions that are easy to understand, from the perspective of what info the main module is operating with (i.e. get_form_id , get_fields .. etc )

File: app.py (hunk 1 above)
Comment (Timtech4u) [reply app.py:24]: This splits the value into two e.g. If I entered `Local*Elections`, the final output is `['Local', 'Elections']`

A way to go around this is to have any user input with * not valid, we could ask the user to use a different symbol if necessary.

What do you suggest?

**Comment by Timtech4u** [comment]
> I'm finding it pretty hard to follow how the app is storing state in Redis. May I suggest creating a separate class/module that is solely responsible of formatting in and parsing out data from Redis?
> 
> It would really make the code cleaner, sp. if that class/module has methods/functions that are easy to understand, from the perspective of what info the main module is operating with (i.e. get_form_id , get_fields .. etc )

Thanks for the feedback, I'll refactor the codes and adopt a module for redis methods.

File: app.py (hunk 2 above)
Comment (Timtech4u) [reply app.py:27]: The initial **ussdTextList** from Africa's Talking has a `''` in it.
Since we're working with the first step which starts at 0, we need to remove the initial `''` so it's initial length i.e. *step* gives us 0, not 1.

File: setup.py (hunk 4 above)
Comment (Timtech4u) [reply setup.py]: Oops, nice catch, I'll add the *redis* package right away.

File: app.py (hunk 3 above)
Comment (Timtech4u) [reply app.py]: Sure, that'll go into my next commit.

File: app.py (hunk 2 above)
Comment (Timtech4u) [reply app.py:27]: To be clear, I made use of *while* (which checks if there's a `''` in the list), not too sure if this adopts iteration, but I'd check :+1:

**Comment by Timtech4u** [comment]
> > I'm finding it pretty hard to follow how the app is storing state in Redis. May I suggest creating a separate class/module that is solely responsible of formatting in and parsing out data from Redis?
> > It would really make the code cleaner, sp. if that class/module has methods/functions that are easy to understand, from the perspective of what info the main module is operating with (i.e. get_form_id , get_fields .. etc )
> 
> Thanks for the feedback, I'll refactor the codes and adopt a module for redis methods.

@tuxpiper  could you confirm if you're able to easily read through the Redis integration with my latest commit.

File: app.py (hunk 2 above)
Comment (tuxpiper) [reply app.py:27]: oh yep you are right :+1: my bad, brain is still coming back from vacation

File: app.py (hunk 1 above)
Comment (tuxpiper) [reply app.py:24]: It would be so much better if Africa's Talking did something with those asterisks ... 

How would we even know that the user has entered an '*' ?  Anyway, this probably deserves its own issue.

File: app.py [hunk 5]
```diff
@@ -1,91 +1,135 @@
...
-                        for i, option in enumerate(options):
-                            response += "\n {} - {}".format(i+1, option)
 
-        return response
+            # Check if we have exhausted Survey **fields**
+            if len(fields) == index:
+                response = "END Thanks for submitting your response."
+                # Post USSD reponses for Survey fields input
+                post_ussd_responses(survey_id, fields, validUserInput[2:])
+
```
Comment (tuxpiper) [inline app.py:136]: how about removing unnecessary entries from Redis so that they don't accummulate? Would that be done somewhere here?

**Review by tuxpiper — CHANGES_REQUESTED** [review]
Thanks! The latest changes do help. I just have one issue about cleaning up redis entries when they are not needed anymore.

File: utils.py [hunk 6]
```diff
@@ -25,4 +25,11 @@ def get_location_coordinates(location):
     if location.address:
         return [location.latitude, location.longitude]
     else:
-        return []
\ No newline at end of file
+        return []
+
+def validate_input(value, type):
```
Comment (rowasc) [inline utils.py:31]: just adding the note about looking into https://pypi.org/project/validation/ as mentioned in the call

File: .gitignore [hunk 7]
```diff
@@ -1,3 +1,4 @@
 venv
 *.pyc
 .env
+redis-play.py
```
Comment (rowasc) [inline .gitignore]: oooo what's this? is it like a configuration file for Redis?

**Comment by tuxpiper** [comment]
@rowasc did you get a chance to review?

**Review by rowasc — COMMENTED** [review]
so sorry I forgot to submit it , @tuxpiper @Timtech4u only a couple questions

File: app.py (hunk 5 above)
Comment (Timtech4u) [reply app.py:136]: Hi @tuxpiper , yes, I'd delete the key in my next commit.

File: .gitignore (hunk 7 above)
Comment (Timtech4u) [reply .gitignore]: This is a test file I used when learning Redis. Thanks for the catch.

**Review by tuxpiper — APPROVED** [review]
Looking good to me! :+1:

**Review by rowasc — APPROVED** [review]
🎉

**Comment by Timtech4u** [comment]
Thanks @rowasc  @tuxpiper
